src/viperleed/calc/files/parameters/file_reader.py

```diff
@@ -0,0 +1,61 @@
+"""Module file_reader of viperleed.calc.classes.
+
+This module is a more abstract version of a settings file reader. It is based on
+@michele-riva's work on the PARAMETERS file reader, which in turn is based on
+the old readPARAMETERS from @fkraushofer.
+
+Defines context-manager, iterator classes for reading parameters from a file.
+"""
+
+__authors__ = (
+    'Alexander M. Imre (@amimre)',
+    'Michele Riva (@michele-riva)',
+    )
+__copyright__ = 'Copyright (c) 2019-2024 ViPErLEED developers'
+__created__ = '2024-10-03'
+__license__ = 'GPLv3+'
+
+from abc import abstractmethod
+from collections.abc import Iterator
+from contextlib import AbstractContextManager
+from pathlib import Path
+
+
+class SettingsFileReader(AbstractContextManager, Iterator):
+    """A context manager that iterates the contents of a settings file
+    (e.g PARAMETERS)."""
+
+    def __init__(self, filename, noisy=True):                                   # TODO: it would be nice to support passing file contents via a StringIO or similar
+        """Initialize instance.
+
+        Parameters
+        ----------
+        filename : str or Path
+            Path to the file to be read.
+        noisy : bool, optional
+            Whether the reader will emit logging messages and raise
+            errors if unknown or malformed parameters are encountered.
+        """
+        self.filename = Path(filename)
+        self.noisy = noisy
+        self._file_obj = None
+        self._current_line = 0
+
+    def __enter__(self):
+        """Enter context."""
+        self._file_obj = self.filename.open('r', encoding='utf-8')
+        self._current_line = 0
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Close file, then return control to caller to handle exceptions."""
+        try:
+            self._file_obj.close()
+        except AttributeError:
+            pass
+        return super().__exit__(exc_type, exc_value, traceback)
+
+    @abstractmethod
+    def __next__(self):
```

Comment:
Could we perhaps provide a simpler interface for subclasses? An example that comes to mind would be an `@abstractmethod` that takes a line as a string and produces information from it [e.g., `def _read_one_line(self, line)`]. Then we could avoid having `__next__` as an abstract one and rather produce a general-enough implementation. It could simply do the job if iterating through the file and return whatever `_read_one_line` does.

To have a general enough implementation, we would then only need to provide a good recipe on how to handle special cases. One case I can see in the PARAMETERS is skipping an 'invalid' line. It would be easy to provide a decent interface: _read_one_line could raise a dedicated exception (e.g., `ShouldSkipLineError`, defined in this module). Then a general enough implementation of `__next__` could be:

```python
    def __next__(self):
        """Return the next understandable information in the file."""
        for line in self._file_obj:
            try:
                return self._read_one_line(line)
            except ShouldSkipLineError:
                continue
        raise StopIteration

    @abstractmethod
    def _read_one_line(self, line):
        """Return understandable information from `line`.

        This method is guaranteed to be called once on each
        line read from the file.

        Parameters
        -------------
        line : str
            A single line read from `self._file_obj.`

        Returns
        --------
        info : object
            Understandable information read from `line`. This is the
            same object returned whenever this reader is iterated over.

        Raises
        -------
        ShouldSkipLineError
            If `line` does not contain any valuable information that is
            worth returning while iterating over the file.
        """
```